blobstore.wit.md

```diff
@@ -0,0 +1,178 @@
+```
+
+// # wasi-cloud blobstore service definition
+//
+// A blobstore is service that can mange containers and objects.
+//
+// ## Object
+// An object is a named sequence of bytes. Objects can be read or written with a stream interface.
+// Once an object is written its contents cannot be modified.
+// Objects can be any size from 1 byte up to the limits imposed by the underlying store.
+// Object names are unique within their container.
+//
+// ## Container
+// A container is a named collection of objects. Container names can be any utf-8 string.
+// Within any instance of the blobstore interface, all containers share the same "flat" namespace.
+
+
+// wasi-cloud Blobstore service definition
+interface "wasi:blob/blobstore" {
+
+  use wasi:blob/types::{ Error, container-metadata, container-name, object-id }
+
+  // creates a new empty container
+  create-container: func(name: container-name) -> result<container,Error>
+
+  // retrieves a container by name
+  get-container: func(name: container-name) -> result<container, Error>
+
+  // deletes a container and all objects within it
+  delete-container: func(name: container-name) -> result<_, Error>
```

Comment:
In other related WASI proposals (i.e., wasi-keyvalue, and wasi-messaging), we've focused uniquely on data plane operations over anything related to the control plane — Do you think these fit as data plane operations?

Reply:
I'm not sure a user of this api would make that distinction. If it were backed by a file system, for example, where containers are folders or directories, and blobs are files, an application might get the folder, then enumerate the files all in one operation. Or say you are writing log files, and create a new folder each day and put the day's log files in that folder.  In those use cases, I think of folders as data.

There are use cases for applying different rights and permissions to containers. For example, a particular application might have rights to read or write files, but not rights to create or delete folders. I imagine that the permissions would be applied as part of the api for obtaining a resource, and a permission failure would result in an error return from one of these operations. Maybe setting permissions could be part of control plane operations, but to me the basic read/write and enumerate operations are more data than control.